memory.c

```diff
@@ -0,0 +1,329 @@
+        }
+    }
+    TracePrintf(0, "allocate_frame: ERROR: No free physical frames available\n");
+    return -1; // No free frames
+}
+
+
+void free_frame(int pfn) {
+    // Basic validation for the physical frame number
+    if (pfn < 0 || pfn >= NUM_VPN) { // Assuming NUM_VPN is the total number of physical frames
+        TracePrintf(0, "free_frame: ERROR: Invalid physical frame number %d\n", pfn);
+        return;
+    }
+
+    // Check if the frame was actually allocated before freeing
+    if (frame_bitMap[pfn] == 0) {
+        TracePrintf(0, "free_frame: WARNING: Attempted to free an already free frame %d\n", pfn);
+        return;
+    }
+
+    frame_bitMap[pfn] = 0; // Mark the frame as free
+    TracePrintf(5, "free_frame: Freed frame %d\n", pfn);
+}
+
+int SetKernelBrk(void *addr) {
+    if (!vm_enabled) {
+        // Before VM enabled: just track the kernel break
+        if (addr > kernel_brk && addr < (void *)KERNEL_STACK_BASE) {
+            kernel_brk = addr;
+            return 0;
+        } else {
+            return ERROR;
+        }
+    } else {
+        // After VM enabled: need to actually allocate frames and update page tables
+        void *old_brk = kernel_brk;
+        void *new_brk = addr;
+
+        // Round up to page boundary
+        new_brk = (void *)UP_TO_PAGE((unsigned int)new_brk);
+
+        // If decreasing break, just update the break pointer
+        if (new_brk <= old_brk) {
+            kernel_brk = new_brk;
+            return 0;
+        }
+
+        // If increasing break, map new frames
+        for (void *p = old_brk; p < new_brk; p += PAGESIZE) {
+            int vpn = (int)p >> PAGESHIFT;
+            int pfn = allocate_frame(); // Get a free physical frame
+
+            if (pfn == ERROR) {
+                return ERROR; // Out of physical memory
+            }
+
+            // Update the page table entry using explicit pointer arithmetic
+            pte_t *current_pte = region0_pt + vpn; // Pointer to the specific PTE
+            current_pte->valid = 1;
+            current_pte->prot = PROT_READ | PROT_WRITE;
+            current_pte->pfn = pfn;
+
+            // Flush the TLB for this page
+            WriteRegister(REG_TLB_FLUSH, (unsigned int)p);
+        }
+
+        kernel_brk = new_brk;
+        return 0;
+    }
+}
+
+
+
+void init_page_table(int kernel_text_start, int kernel_data_start, int kernel_brk_start) {
+    TracePrintf(0, "Initializing page table...\n");
+    // Set the base address for the region0 page table in hardware
+    unsigned int last_physical_address = MAX_PMEM_SIZE - 1; // Calculate the last physical address
+    unsigned int page_table_size = MAX_PT_LEN * sizeof(struct pte); // Calculate size needed for the page table
+    unsigned int p_address_region0_base = DOWN_TO_PAGE(last_physical_address - page_table_size); // Place page table at the end of physical memory
+    region0_pt = (pte_t *)p_address_region0_base; // Cast to pte_t*

```

Comment:
it won't work to put the page table at the top of physical memory, because the physical memory size is larger than that of virtual memory. Suppose there is 0x2000 in each VM region (so R0 goes 0x0000 to 0x1999 and R1 goes 0x2000 to 0x3999) and that theres 0x10000 of physical memory. if you place the page table at 0x09000, it'll be accessible before VM is enabled -- but you can't set it up so that the virtual address 0x09000 points to physical 0x09000, since virtual memory doesn't go that high. 

Instead, I recommend placing the entire region0_pt within the data region of your kernel by initializing it (on line 14) as `pte_t region0_pt[MAX_PT_LEN];` this will be between kernel data start and kernel orig brk, and everything will be happier.